aten/src/ATen/native/BatchLinearAlgebraKernel.cpp

```diff
@@ -161,6 +163,102 @@ std::tuple<Tensor, Tensor> eig_kernel_impl(const Tensor& self, bool& eigenvector
   return std::tuple<Tensor, Tensor>(vals_, vecs_);
 }
 
+/*
+  Computes eigenvalues and eigenvectors of the input that is stored initially in 'vectors'.
+  The computation is done in-place: 'vectors' stores the input and will be overwritten,
+  'values' should be an allocated empty array.
+  'infos' is used to store information for possible checks for error.
+  'upper' controls the portion of input matrix to consider in computations
+  'compute_eigenvectors' controls whether eigenvectors should be computed.
+  This function doesn't do any error checks and it's assumed that every argument is valid.
+*/
+template <typename scalar_t>
+void apply_lapack_eigh(Tensor& values, Tensor& vectors, Tensor& infos, bool upper, bool compute_eigenvectors) {
+#ifndef USE_LAPACK
+  TORCH_CHECK(
+      false,
+      "Calling torch.linalg.eigh/eigvalsh on a CPU tensor requires compiling ",
+      "PyTorch with LAPACK. Please use PyTorch built with LAPACK support.");
+#else
+  using value_t = typename c10::scalar_value_type<scalar_t>::type;
+
+  // NumPy allows lowercase input for UPLO argument
+  // It is assumed that uplo_str is either "U" or "L"
+  char uplo = upper ? 'U' : 'L';
+  char jobz = compute_eigenvectors ? 'V' : 'N';
+
+  auto n = vectors.size(-1);
+  auto lda = std::max<int64_t>(1, n);
+  auto batch_size = batchCount(vectors);
+
+  auto vectors_stride = matrixStride(vectors);
+  auto values_stride = values.size(-1);
+
+  auto vectors_data = vectors.data_ptr<scalar_t>();
+  auto values_data = values.data_ptr<value_t>();
+  auto infos_data = infos.data_ptr<int>();
+
+  // Using 'int' instead of int32_t or int64_t is consistent with the current LAPACK interface
+  // It really should be changed in the future to something like lapack_int that depends on the specific LAPACK library that is linked
+  // or switch to supporting only 64-bit indexing by default.
+  int lwork = -1;
+  int lrwork = -1;
+  int liwork = -1;
+  scalar_t lwork_query;
+  value_t rwork_query;
+  int iwork_query;
+
+  // call lapackSyevd once to get the optimal size for work data
+  scalar_t work_query;
+  lapackSyevd<scalar_t, value_t>(jobz, uplo, n, vectors_data, lda, values_data,
+    &lwork_query, lwork, &rwork_query, lrwork, &iwork_query, liwork, infos_data);
+
+  lwork = std::max<int>(1, real_impl<scalar_t, value_t>(lwork_query));
+  Tensor work = at::empty({lwork}, vectors.options());
+  auto work_data = work.data_ptr<scalar_t>();
+
+  liwork = std::max<int>(1, iwork_query);
+  Tensor iwork = at::empty({liwork}, vectors.options().dtype(at::kInt));
+  auto iwork_data = iwork.data_ptr<int>();
+
+  Tensor rwork;
+  value_t* rwork_data = nullptr;
+  if (vectors.is_complex()) {
+    lrwork = std::max<int>(1, rwork_query);
+    rwork = at::empty({lrwork}, values.options());
+    rwork_data = rwork.data_ptr<value_t>();
+  }
+
+  // Now call lapackSyevd for each matrix in the batched input
+  for (const auto i : c10::irange(batch_size)) {
```

Comment:
Could we skip the first matrix since it's already computed above to get the work size? Or is there another way to figure out the size?

Reply:
No actual computation of eigendecomposition is done during the above call to `lapackSyevd`. LAPACK routines work such that when worksize arguments are -1, then only optimal work sizes are computed and that's it. That can be confusing that's why cuSOLVER has separate functions with `_bufferSize` ending, and the new Intel's oneAPI MKL has `_scratchpad_size` functions for that.

Reply:
Thanks for clarifying.